lib/ledenet/api.rb

```diff
@@ -14,56 +14,112 @@ def initialize(device_address, options = {})
+      end
     end
 
-    def update_color(r, g, b)
-      checksum = color_checksum(r, g, b)
+    def update_color(r,g,b) # Supports legacy color Updates
+      update_ufo(r,g,b,0,false)
+      true
+    end
 
-      send_bytes_action(0x31, r, g, b, 0xFF, 0, 0x0F, checksum)
+    def update_ufo(r,g,b,w,persist) # Update a UFO wireless device
+      msg = Array.new
+      if persist
+        msg << 0x31
+      else
+        msg << 0x41 << r << g << b << w << 0x00 << 0x0f
+      end
+      checksum = calc_checksum(msg)
+      send_bytes_action(*msg, checksum)
+      true
+    end
 
+    def update_bulb_color(r, g, b, persist) # Update a Bulb wireless device's color
+      msg = Array.new
+      if persist
+        msg << 0x31
+      else
+        msg << 0x41 << r << g << b << 0x00 << 0xf0 << 0x0f
+      end
+      checksum = calc_checksum(msg)
+      send_bytes_action(*msg, checksum)
       true
     end
 
-    def current_color
-      status.bytes[6, 3]
+    def update_bulb_white(w, persist) # Update a Bulb wireless device's WW level
+    msg = Array.new
+    if persist
+        msg << 0x31
+      else
+        msg << 0x41 << 0x00 << 0x00 << 0x00 << w << 0x0f << 0x0f
+      end
+        checksum = calc_checksum(msg)
+        send_bytes_action(*msg, checksum)
+        true
+    end
+
+    def current_status # Gets bytes 6-9 and returns them as Integers from 0-255 (Red,Green,Blue, and WW) and return power status as string
+      current_packet = Array.new
+      current_packet = status
+      power_state = "off"
+      power_state = "on" if current_packet[2].unpack('C').to_s.delete('[]') == "35"
+      return Integer(current_packet[6].unpack('C').to_s.delete('[]')).to_i,Integer(current_packet[7].unpack('C').to_s.delete('[]')).to_i,Integer(current_packet[8].unpack('C').to_s.delete('[]')).to_i,Integer(current_packet[9].unpack('C').to_s.delete('[]')).to_i, power_state
     end
 
+    def current_color
+      current_packet = Array.new
+      current_packet = status
+      currentColors = Array.new
+      currentColors << Integer(current_packet[6].unpack('C').to_s.delete('[]')).to_i << Integer(current_packet[7].unpack('C').to_s.delete('[]')).to_i << Integer(current_packet[8].unpack('C').to_s.delete('[]')).to_i
+    end
+    
     def reconnect!
       create_socket
     end
 
+    def getInfo
+      msg = Array.new
+      current_packet = Array.new
+      current_packet = status
+      msg << Integer(current_packet.each { |x| puts x }.unpack('C').to_s.delete('[]'))
+    end
+
     private
-      def color_checksum(r, g, b)
-        (r + g + b + 0x3F) % 0x100
+      def calc_checksum(bytes)
+        bytes.inject{|sum,x| sum + x } % 0x100
       end
 
       def status
         socket_action do
-          send_bytes(0x81, 0x8A, 0x8B, 0x96)
-
-          # Example response:
-          # [129, 4, 35, 97, 33, 9, 11, 22, 33, 255, 3, 0, 0, 119]
-          #                         R   G   B   WW            ^--- LSB indicates on/off
+          msg = Array.new
```

Comment:
To be frank, I'm not sure why I initialized arrays the way I did. Thank you for your notes on all this—I'll edit and clean things up with your info and file a re-request.